src/Features/Core/Portable/Completion/Providers/Snippets/AbstractSnippetCompletionProvider.cs

```diff
@@ -0,0 +1,119 @@
+            var strippedText = await strippedDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var allChangesText = strippedText.WithChanges(snippet.TextChanges);
+            var allChangesDocument = document.WithText(allChangesText);
+            var allTextChanges = await allChangesDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var change = Utilities.Collapse(allChangesText, allTextChanges.AsImmutable());
+            return CompletionChange.Create(change, newPosition: snippet.CursorPosition, includesCommitCharacter: true);
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var position = context.Position;
+            var service = document.GetLanguageService<ISnippetService>();
+
+            if (service == null)
+            {
+                return;
+            }
+
+            var (strippedDocument, newPosition) = await GetDocumentWithoutInvocationAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            var snippets = await service.GetSnippetsAsync(strippedDocument, newPosition, cancellationToken).ConfigureAwait(false);
+
+            foreach (var snippetData in snippets)
+            {
+                if (snippetData is null)
+                {
+                    continue;
+                }
+
+                var snippetValue = snippetData.Value;
+                var completionItem = SnippetCompletionItem.Create(
+                    displayText: snippetValue.DisplayName,
+                    displayTextSuffix: "",
+                    position: position,
+                    glyph: Glyph.Snippet);
+                context.AddItem(completionItem);
+            }
+        }
+
+        /// Gets the document without whatever text was used to invoke the completion.
+        /// Also gets the new position the cursor will be on.
+        /// Returns the original document and position if completion was invoked using Ctrl-Space.
+        /// 
+        /// public void Method()
+        /// {
+        ///     $$               //invoked by typing Ctrl-Space
+        /// }
+        /// Example invoking when span is not empty:
+        /// public void Method()
+        /// {
+        ///     Wr$$             //invoked by typing out the completion 
+        /// }
+        private static async Task<(Document, int)> GetDocumentWithoutInvocationAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var originalText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var (startPosition, endPosition) = GetStartAndEndPositionOfInvocationText(originalText, position);
```

Comment:
using the term 'invocation' is a bit confusing as invocation means `Foo()` in C#/VB.  I would just say `GetSpanOfAlreadyWrittenWord`